src/transformers/parallelization_utils.py

```diff
@@ -0,0 +1,1117 @@
+                world_size=world_size,
+                gpu_index=gpu_index,
+                slice_bias=False,
+            )
+            for layer in layers
+        ]
+
+    def _process_additional_layers(
+        self,
+        model,
+        mpu,
+        additional_layers,
+        config,
+    ):
+        additional_layers += self.policy.modelwise_copy_to_all(model, config)
+
+        for layer in additional_layers:
+            if layer.weight is not None:
+                layer.weight.data = layer.weight.to(self.device)
+            if layer.bias is not None:
+                layer.bias.data = layer.bias.to(self.device)
+            self._postprocess(layer, {"mpu": mpu})
+
+    def _postprocess(self, layer, attributes):
+        if layer.replace is not None:
+            for k, v in layer.replace.items():
+                for k_, v_ in attributes.items():
+                    setattr(k, k_, v_)
+                k.__class__ = v
+
+        if self.hold_params:
+            if layer.name in self.param_dict:
+                self.param_dict[layer.name].append(layer)
+            else:
+                self.param_dict[layer.name] = [layer]
+
+
+class ParallelizationMixin(object):
+    def _get_default_model_object(self):
+        raise NotImplementedError
+
+    def _get_parallelism_policy(self):
+        raise NotImplementedError
+
+    def _get_head_layers(self):
+        return []
+
+    @classmethod
+    def from_pretrained_with_parallel(
+        cls,
+        pretrained_model_name_or_path: Optional[Union[str, os.PathLike]],
+        tensor_model_parallel_size: int = 1,
+        pipeline_model_parallel_size: int = 1,
+        fp16: bool = False,
+        vocab_parallel_embedding: bool = None,
+        *model_args,
+        **kwargs,
+    ):
+        model = cls.from_pretrained(pretrained_model_name_or_path, *model_args, **kwargs)
+
+        if fp16 is True:
+            model = model.half()
+
+        assert pipeline_model_parallel_size == 1, "Currently, We only support tensor model parallelism."
+
+        if vocab_parallel_embedding is None:
+            if hasattr(model.config, "vocab_parallel_embedding"):
+                vocab_parallel_embedding = model.config.vocab_parallel_embedding
+
+        mpu = MPU(
+            tensor_model_parallel_size=tensor_model_parallel_size,
+            pipeline_model_parallel_size=pipeline_model_parallel_size,
+        )
+
+        engine = ParallelEngine(
+            hold_params=False,
+            skip_bias_add=False,
+            policy=model._get_parallelism_policy(),
+            vocab_parallel_embedding=vocab_parallel_embedding,
+        )
+
+        engine.parallelize(
+            model=model._get_default_model_object(),
+            mpu=mpu,
+        )
+
+        setattr(model, "mpu", mpu)
+        # This allows the Trainer to call the MPU for data + model parallelism
+        # example `ddp = DistributedDataParallel(..., process_group=model.mpu.get_data_parallel_group())`
+
```

Comment:
We need to change the Trainer code for 2D parallelism. You can prevent the use of naive DP, and input the mpu to distributed data parallel launcher like `deepspeed.initialize`, `nn.parallel.DistributedDataParallel`, `FullyShardedDDP`, etc.